runtime/tests/org.eclipse.core.tests.runtime/src/org/eclipse/core/tests/runtime/jobs/JobGroupTest.java

```diff
@@ -1293,6 +1309,71 @@ public IStatus run(IProgressMonitor monitor) {
 		assertTrue("2.0", numShouldCancelCalled[0] < NUM_JOBS_LIMIT + 10);
 	}
 
+	/**
+	 * Tested scenario: - 2 jobs are started in parallel and the first one returns a
+	 * CANCEL_STATUS while the second one is still running. This lets
+	 * <code>JobGroup.shouldCancel</code> return <code>true</code> which triggers
+	 * the cancelation of any running job.
+	 *
+	 * Expected result: The second job is explicitly canceled
+	 *
+	 */
+	public void testShouldCancel_5() {
+		// the job group allows for 2 threads so both jobs will be started in
+		// parallel
+		JobGroup jobGroup = new JobGroup("JobGroup", 2, 2) {
+			@Override
+			protected boolean shouldCancel(IStatus lastCompletedJobResult, int numberOfFailedJobs,
+					int numberOfCanceledJobs) {
+				// cancel the group as soon as one job in it returns the CANCEL_STATUS
+				return numberOfCanceledJobs > 0;
+			}
+		};
+
+		TestBarrier2 returnsCancelStatusBarrier = new TestBarrier2();
+		TestBarrier2 canceledJobBarrier = new TestBarrier2();
+
+		Job returnsCancelStatusJob = Job.create("The one that cancels", __ -> {
+			// This job running. Wait until the other job is running too.
+			returnsCancelStatusBarrier.setStatus(TestBarrier2.STATUS_RUNNING);
+			canceledJobBarrier.waitForStatus(TestBarrier2.STATUS_RUNNING);
+
+			return Status.CANCEL_STATUS;
+		});
+
+		Job canceledJob = new Job("The one that is explicitly canceled") {
+			@Override
+			protected IStatus run(IProgressMonitor monitor) {
+				// This job running. Wait until the other job is running too.
+				canceledJobBarrier.setStatus(TestBarrier2.STATUS_RUNNING);
+				returnsCancelStatusBarrier.waitForStatus(TestBarrier2.STATUS_RUNNING);
+
+				// give it time: this Job needs to be running at the time the other job finishes
+				canceledJobBarrier.waitForStatus(TestBarrier2.STATUS_DONE);
+				return Status.OK_STATUS;
+			}
+
+			@Override
+			protected void canceling() {
+				canceledJobBarrier.setStatus(TestBarrier2.STATUS_DONE);
+			}
+		};
+
+		returnsCancelStatusJob.setJobGroup(jobGroup);
+		returnsCancelStatusJob.schedule();
+
+		canceledJob.setJobGroup(jobGroup);
+		canceledJob.schedule();
+
+		waitForCompletion(jobGroup);
+
+		/*
+		 * The fact that we reached the end of the method means that the 2nd job was
+		 * successfully canceled i.e. that the test passed. This is why there are no
+		 * assertions in this test.
```

Comment:
This comment is a bit misleading, as `TestBarrier2::waitForStatus` actually is an assertion (although the name does not express that).